### PR TITLE
Support smart completion/squeeze some bugs 

### PR DESCRIPTION
**Pull request by xqliu**
1. Support smart completion(the method whose return value is same as the
   variable or field to be assigned will be shown before other candidates
   in auto completion menu)
2. Add netbeans config folder to .gitignore

Note: For detail feature description and behaviour of this implementation, please refer to 

https://github.com/mopemope/meghanada-emacs/pull/102

Partially addressed #51 

For detail information please refer to #51

### CODE REVIEW

File: server/src/main/java/meghanada/completion/JavaCompletion.java [hunk 1]
```diff
@@ -37,10 +38,7 @@
 import meghanada.config.Config;
 import meghanada.index.IndexDatabase;
 import meghanada.project.Project;
-import meghanada.reflect.CandidateUnit;
-import meghanada.reflect.ClassIndex;
-import meghanada.reflect.FieldDescriptor;
-import meghanada.reflect.MemberDescriptor;
+import meghanada.reflect.*;
```
Comment (mopemope) [inline server/src/main/java/meghanada/completion/JavaCompletion.java]: No wildcard imports.

File: server/src/main/java/meghanada/completion/JavaCompletion.java [hunk 2]
```diff
@@ -237,7 +235,6 @@ private static CompletionMatcher createClassCompletionMatcher(
     final CompletionMatcher classMatcher = getClassCompletionMatcher(prefix);
     // add this member
     completionThisMembers(typeScope, prefix, result, matcher);
-    completionThisMembers(prefix, result, fqcn);
```
Comment (mopemope) [inline server/src/main/java/meghanada/completion/JavaCompletion.java]: Why did you delete it ?

File: server/src/main/java/meghanada/completion/JavaCompletion.java [hunk 3]
```diff
@@ -739,20 +754,99 @@ private Source getSource(final File file) throws IOException, ExecutionException
...
+            typeOrMember = entry.getValue().fqcn;
+            found = true;
+            break;
+          }
+        }
+        if (!found) {
+          Collection<FieldAccess> fields = source.getFieldAccesses();
+          for (FieldAccess field : fields) {
+            if (field.name.equals(typeOrMember)) {
+              found = true;
```
Comment (mopemope) [inline server/src/main/java/meghanada/completion/JavaCompletion.java]: `found` is never used.

File: server/src/main/java/meghanada/completion/JavaCompletion.java [hunk 4]
```diff
@@ -739,20 +754,99 @@ private Source getSource(final File file) throws IOException, ExecutionException
...
+              typeOrMember = field.returnType;
+            }
+          }
+        }
+      } catch (Exception ignored) {
```
Comment (mopemope) [inline server/src/main/java/meghanada/completion/JavaCompletion.java]: Add `log.catching`

File: server/src/main/java/meghanada/completion/JavaCompletion.java (hunk 2 above)
Comment (xqliu) [reply server/src/main/java/meghanada/completion/JavaCompletion.java]: Dear,

I found that when completion in current class, for example,  take the follow code for example,

```
class A {
   public void aMethod() {}

  public void bMethod() {
    this. //<-- Cursor here
  }
}
```

aMethod will popup for two times, and after I deleted the above line, it only appears for one time.

**Comment by xqliu** [comment]
Dear, 

I have added more test cases to JavaCompletionTest.java for smart completion.

Please help to review it.

Thanks :)

File: server/src/main/java/meghanada/completion/JavaCompletion.java (hunk 2 above)
Comment (mopemope) [reply server/src/main/java/meghanada/completion/JavaCompletion.java]: I see. OK.

**Comment by mopemope** [comment]
@xqliu Merged. Thanks
